### PR TITLE
Add text regex support for Event watcher 

### LINKED ISSUE #2212: Empower EventWatcher to replace a YAML file that holds separators
**What would you like to be added**:
It's not uncommon that you want to replace YAML files that are already merged with `---`, the end of directives marker.

```yaml
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      containers:
      - name: foo
        image: foo:v0.8.0

---
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      containers:
      - name: bar
        image: bar:v0.8.0
```

In above example, `$.spec.template.spec.containers[0].image` exists in both docs but we (in some case) only want to update the second doc.

**Why is this needed**:

### LINKED ISSUE #2341: Event watcher can't update YAML files including templating syntax
**What would you like to be added**:
Let's say you update a YAML file like below by Event watcher

```yaml
apiVersion: pipecd.dev/v1beta1
kind: LambdaFunction
spec:
  image: foo
  environments:
    FOO: {{ .encryptedSecrets.foo }}
```

Event watcher never updates because the file includes `{{}}`.

```
failed to commit outdated files	{"error": "failed to get value at $.spec.image in /tmp/event-watcher621282109/repo524333230/tmp-repo/services/foo/bar/function.yaml: [6:11] unexpected mapping key\n       3 | spec:\n       4 |   image: foo\n       5 |   environments:\n    >  6 |     FOO: {{ .encryptedSecrets.foo }}\n                     ^"}
```

It's common to have a templating syntax for [Encrypted secret](https://pipecd.dev/docs/user-guide/secret-management/) so it should be handled as soon as possible.

**Why is this needed**:

### LINKED ISSUE #2345: Want Event watcher to replace a part of YAML field
**What would you like to be added**:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: playground
spec:
  template:
    spec:
      containers:
      - name: helloworld
        image: gcr.io/pipecd/helloworld:{v0.8.0} # <- it's useful if it could replace only v0.8.0
```

**Why is this needed**:

### PR DESCRIPTION
**Pull request by nakabonne**
**What this PR does / why we need it**:
This PR adds the implementation of https://github.com/pipe-cd/pipe/issues/2346#issuecomment-897381566

**Which issue(s) this PR fixes**:

Fixes https://github.com/pipe-cd/pipe/issues/2346
Fixes #2212 
Fixes #2341
Fixes #2345

**Does this PR introduce a user-facing change?**:

```release-note
Add text regex support for Event watcher
```

### CODE REVIEW

**Comment by nakabonne** [comment]
We settled on employing https://github.com/pipe-cd/pipe/issues/2346#issuecomment-898178202
/hold

File: pkg/config/event_watcher.go [hunk 1]
```diff
@@ -50,6 +50,10 @@ type EventWatcherReplacement struct {
 	JSONField string `json:"jsonField"`
 	// The HCL path to the field to be updated.
 	HCLField string `json:"HCLField"`
+	// The regex string that specify what should be replaced.
+	// The only first capturing group enclosed by `()` will be replaced with the new value.
```
Comment (khanhtc1202) [inline pkg/config/event_watcher.go:54]:
```suggestion
	// The regex string specifying what should be replaced.
	// Only the first capturing group enclosed by `()` will be replaced with the new value.
```
nits

File: pkg/app/piped/eventwatcher/eventwatcher.go [hunk 2]
```diff
@@ -320,3 +324,51 @@ func convertStr(value interface{}) (out string, err error) {
 	}
 	return
 }
+
```
Comment (khanhtc1202) [inline pkg/app/piped/eventwatcher/eventwatcher.go:327]: nits, should we add a comment for this just as `modifyYAML` function has? It's a nits and it's up to you 🙆‍♀️

File: pkg/app/piped/eventwatcher/eventwatcher.go [hunk 3]
```diff
@@ -320,3 +324,51 @@ func convertStr(value interface{}) (out string, err error) {
...
+func modifyText(path, regexText, newValue string) ([]byte, bool, error) {
```
[the remark this answers is not in the document]
Comment (nakabonne) [reply pkg/app/piped/eventwatcher/eventwatcher.go:328]: That's more helpful 👍

**Comment by nakabonne** [comment]
Applied, PTAL 😄

**Comment by nakabonne** [comment]
I'm looking to add a couple of examples after this got mereged.

**Comment by nghialv** [comment]
Nice work!
/lgtm

**Review by khanhtc1202 — APPROVED** [review]

**Pull request merged**